No edges left after adaptive sampling

### ✅ Code of conduct checklist

- [x] 🌱 I am using the **_latest version_** of the [code](https://github.com/drprojects/superpoint_transformer/tree/master).
- [x] 👩‍💻 I made sure the bug concerns the official [project's codebase](https://github.com/drprojects/superpoint_transformer/tree/master) and not something I coded myself on top of it. (We only provide support for code we wrote and released ourselves).
- [x] 🔎 I took appropriate **_time_** to investigate the problem before filing an issue, but am unable to solve it myself.
- [x] 📙 I **_thoroughly_** went through the [README](https://github.com/drprojects/superpoint_transformer/blob/master/README.md), but could not find the solution there.
- [x] 📘 I **_thoroughly_** went through the tutorial [slides](media/superpoint_transformer_tutorial.pdf), [notebook](notebooks/superpoint_transformer_tutorial.ipynb), and [video](https://www.youtube.com/watch?v=2qKhpQs9gJw), but could not find the solution there.
- [x] 📗 I **_thoroughly_** went through the [documentation](https://github.com/drprojects/superpoint_transformer/tree/master/docs), but could not find the solution there.
- [x] 📜 I went through the **_docstrings_** and **_comments_** in the [source code](https://github.com/drprojects/superpoint_transformer/tree/master) parts relevant to my problem, but could not find the solution there.
- [x] 👩‍🔧 I searched for [**_similar issues_**](https://github.com/drprojects/superpoint_transformer/issues), but could not find the solution there.
- [x] 🔎 I made sure my bug is related to the [project's codebase](https://github.com/drprojects/superpoint_transformer/tree/master) and is not in fact a sub-dependency issue (e.g.: FRNN installation issue, PyTorch installation issue, PyTorch Geometric not installed with GPU support, ...).
- [x] ⭐ Since I am showing interest in the project, I took the time to give the [repo](https://github.com/drprojects/superpoint_transformer/tree/master) a ⭐ to show support. **Please do, it means a lot to us !**

### 🐛 Describe the bug

I am using ezsp and I am trying to do some training using my own data (lidar similar to dales, just modifed the TILES to use my own files) to create a partition checkpoint, it works for a while but then after n (22, 42, etc, including after a val) epochs I get an assertion error from:

https://github.com/drprojects/superpoint_transformer/blob/69e401d1fc5419e6e6be24615925892a2f7a53ca/src/loss/partition_criterion.py#L110-L112

Do you have an insight on what might be causing this?

I noticed that the function:https://github.com/drprojects/superpoint_transformer/blob/69e401d1fc5419e6e6be24615925892a2f7a53ca/src/loss/partition_criterion.py#L107-L108

is called with a minority_class argument of 0, should I be using 0 for specific classes e.g. my junk/other class? I currently have powerlines in class 0, would it be better to contain something more common like ground or vegetation as the original dales data does?

```
Epoch 55/199 ━━━━━━━━━━━━━━━━━━━━╸━━━━━━━━━━━━━━━━━━━ 389/742 0:00:13 • 0:00:13 28.43it/s
v_num: bwua train/partition_loss: 0.140 val/partition_loss: 0.242 val/n_sp: 3180382.000 val/omiou: 88.578 val/ooa: 96.832
val/omacc: 93.175 val/oiou_Class 0: 92.662 val/oiou_Class 1: 88.003 val/oiou_Class 2: 93.360 val/oiou_Class 4: 98.456
val/oiou_Class 6: 94.627 val/oiou_Class 7: 73.711 val/omiou_best: 88.578 val/ooa_best: 96.843 val/omacc_best: 93.175
val/points_per_superpoint: 29.928 val/oiou_Class 5: 79.228
```

Is there a way to know which H5 is currently being processed on failure, I may be able to investigate the data myself or forward you the file causing the issue if I can narrow it down

### 📜 Log output

```shell
[2026-02-17 04:58:52,376][src.utils.utils][INFO] - Enforcing tags! <cfg.extras.enforce_tags=True>
[2026-02-17 04:58:52,381][src.utils.utils][INFO] - Printing config tree with Rich! <cfg.extras.print_config=True>
[2026-02-17 04:58:52,925][__main__][INFO] - Instantiating datamodule <src.datamodules.x_datamodule.XDataModule>
[2026-02-17 04:58:53,089][__main__][INFO] - Instantiating model <src.models.semantic.PartitionAndSemanticModule>
[2026-02-17 04:58:53,166][src.models.semantic][INFO] - The CNN is trained from scratch
[2026-02-17 04:58:53,170][__main__][INFO] - Instantiating callbacks...
[2026-02-17 04:58:53,170][src.utils.utils][INFO] - Instantiating callback <pytorch_lightning.callbacks.ModelCheckpoint>
[2026-02-17 04:58:53,174][src.utils.utils][INFO] - Instantiating callback <pytorch_lightning.callbacks.EarlyStopping>
[2026-02-17 04:58:53,175][src.utils.utils][INFO] - Instantiating callback <pytorch_lightning.callbacks.RichModelSummary>
[2026-02-17 04:58:53,176][src.utils.utils][INFO] - Instantiating callback <pytorch_lightning.callbacks.RichProgressBar>
[2026-02-17 04:58:53,176][src.utils.utils][INFO] - Instantiating callback <pytorch_lightning.callbacks.LearningRateMonitor>
[2026-02-17 04:58:53,177][src.utils.utils][INFO] - Instantiating callback <pytorch_lightning.callbacks.GradientAccumulationScheduler>
[2026-02-17 04:58:53,177][__main__][INFO] - Instantiating loggers...
[2026-02-17 04:58:53,178][src.utils.utils][INFO] - Instantiating logger <pytorch_lightning.loggers.wandb.WandbLogger>
[2026-02-17 04:58:53,181][__main__][INFO] - Instantiating trainer <pytorch_lightning.Trainer>
[2026-02-17 04:58:53,198][__main__][INFO] - Logging hyperparameters!
[2026-02-17 04:58:57,321][__main__][INFO] - Starting training!
[2026-02-17 04:58:57,349][src.datasets.base][INFO] - Dataset hash: "338653f76bc39d2786f564ed9b7e9a3b"
[2026-02-17 04:58:57,350][src.datasets.base][INFO] - Preprocessed data can be found at:                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/train/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/val/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/test/338653f76bc39d2786f564ed9b7e9a3b                  
[2026-02-17 04:58:57,359][src.datasets.base][INFO] - Dataset hash: "338653f76bc39d2786f564ed9b7e9a3b"
[2026-02-17 04:58:57,359][src.datasets.base][INFO] - Preprocessed data can be found at:                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/train/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/val/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/test/338653f76bc39d2786f564ed9b7e9a3b                  
[2026-02-17 04:58:57,369][src.datasets.base][INFO] - Dataset hash: "338653f76bc39d2786f564ed9b7e9a3b"
[2026-02-17 04:58:57,369][src.datasets.base][INFO] - Preprocessed data can be found at:                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/train/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/val/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/test/338653f76bc39d2786f564ed9b7e9a3b                  
[2026-02-17 04:58:57,379][src.datasets.base][INFO] - Dataset hash: "338653f76bc39d2786f564ed9b7e9a3b"
[2026-02-17 04:58:57,380][src.datasets.base][INFO] - Preprocessed data can be found at:                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/train/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/val/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/test/338653f76bc39d2786f564ed9b7e9a3b                  
[2026-02-17 04:58:57,405][src.datasets.base][INFO] - Dataset hash: "338653f76bc39d2786f564ed9b7e9a3b"
[2026-02-17 04:58:57,405][src.datasets.base][INFO] - Preprocessed data can be found at:                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/train/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/val/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/test/338653f76bc39d2786f564ed9b7e9a3b                  
[2026-02-17 04:58:57,414][src.datasets.base][INFO] - Dataset hash: "338653f76bc39d2786f564ed9b7e9a3b"
[2026-02-17 04:58:57,415][src.datasets.base][INFO] - Preprocessed data can be found at:                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/train/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/val/338653f76bc39d2786f564ed9b7e9a3b                  
 - /workspace/superpoint_transformer/data/jetxyz/processed/test/338653f76bc39d2786f564ed9b7e9a3b                  
[2026-02-17 05:24:58,269][src.utils.utils][ERROR] - 
Traceback (most recent call last):
  File "/workspace/superpoint_transformer/src/utils/utils.py", line 51, in wrap
    metric_dict, object_dict = task_func(cfg=cfg)
  File "src/train.py", line 115, in train
    trainer.fit(model=model, datamodule=datamodule, ckpt_path=cfg.get("ckpt_path"))
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 538, in fit
    call._call_and_handle_interrupt(
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/trainer/call.py", line 47, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 574, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 981, in _run
    results = self._run_stage()
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1025, in _run_stage
    self.fit_loop.run()
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/loops/fit_loop.py", line 205, in run
    self.advance()
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/loops/fit_loop.py", line 363, in advance
    self.epoch_loop.run(self._data_fetcher)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/loops/training_epoch_loop.py", line 140, in run
    self.advance(data_fetcher)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/loops/training_epoch_loop.py", line 250, in advance
    batch_output = self.automatic_optimization.run(trainer.optimizers[0], batch_idx, kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/loops/optimization/automatic.py", line 190, in run
    self._optimizer_step(batch_idx, closure)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/loops/optimization/automatic.py", line 268, in _optimizer_step
    call._call_lightning_module_hook(
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/trainer/call.py", line 167, in _call_lightning_module_hook
    output = fn(*args, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/core/module.py", line 1306, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/core/optimizer.py", line 153, in step
    step_output = self._strategy.optimizer_step(self._optimizer, closure, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/strategies/strategy.py", line 238, in optimizer_step
    return self.precision_plugin.optimizer_step(optimizer, model=model, closure=closure, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/plugins/precision/precision.py", line 122, in optimizer_step
    return optimizer.step(closure=closure, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/torch/optim/optimizer.py", line 76, in _use_grad
    ret = func(self, *args, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/torch/optim/adamw.py", line 164, in step
    loss = closure()
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/plugins/precision/precision.py", line 108, in _wrap_closure
    closure_result = closure()
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/loops/optimization/automatic.py", line 144, in __call__
    self._result = self.closure(*args, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/loops/optimization/automatic.py", line 129, in closure
    step_output = self._step_fn()
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/loops/optimization/automatic.py", line 317, in _training_step
    training_step_output = call._call_strategy_hook(trainer, "training_step", *kwargs.values())
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/trainer/call.py", line 319, in _call_strategy_hook
    output = fn(*args, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/pytorch_lightning/strategies/strategy.py", line 390, in training_step
    return self.lightning_module.training_step(*args, **kwargs)
  File "/workspace/superpoint_transformer/src/models/semantic.py", line 692, in training_step
    loss, output = self.model_step(batch)
  File "/workspace/superpoint_transformer/src/models/semantic.py", line 1584, in model_step
    loss, partition_output = self.partition_criterion(
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/envs/spt/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/workspace/superpoint_transformer/src/loss/partition_criterion.py", line 59, in forward
    edge_classification_loss, edge_classification_output = self.edge_classification_loss(partition_output)
  File "/workspace/superpoint_transformer/src/loss/partition_criterion.py", line 110, in edge_classification_loss
    assert sampled_indices.numel() > 0, \
AssertionError: No edges left after adaptive sampling. Probably because there were no inter-edges (minority class).
[2026-02-17 05:24:58,274][src.utils.utils][INFO] - Closing loggers...
[2026-02-17 05:24:58,274][src.utils.utils][INFO] - Closing wandb!
```

### 🤖 Steps to reproduce the bug

```bash
PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True python src/train.py experiment=partition/dales_ezsp.yaml
```

### 📚 Additional information

_No response_

### 🖥️ Environment

```shell
Collecting environment information...
PyTorch version: 2.2.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.28.1
Libc version: glibc-2.35

Python version: 3.8.20 (default, Oct  3 2024, 15:24:27)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.14.0-1018-aws-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 12.3.107
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA L4
Nvidia driver version: 580.105.08
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
Is XPU available: N/A
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  16
On-line CPU(s) list:                     0-15
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 7R13 Processor
CPU family:                              25
Model:                                   1
Thread(s) per core:                      2
Core(s) per socket:                      8
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                5299.99
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch topoext ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr rdpru wbnoinvd arat npt nrip_save vaes vpclmulqdq rdpid
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               256 KiB (8 instances)
L1i cache:                               256 KiB (8 instances)
L2 cache:                                4 MiB (8 instances)
L3 cache:                                32 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-15
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Mitigation; Clear CPU buffers
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==8.9.2.26
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-nccl-cu12==2.19.3
[pip3] nvidia-nvjitlink-cu12==12.9.86
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] pytorch-lightning==2.4.0
[pip3] torch==2.2.0+cu121
[pip3] torch_cluster==1.6.3+pt22cu121
[pip3] torch_geometric==2.3.0
[pip3] torch-graph-components==0.1.1
[pip3] torch_ransac3d==2.0.0
[pip3] torch_scatter==2.1.2+pt22cu121
[pip3] torch-tb-profiler==0.4.3
[pip3] torchmetrics==0.11.4
[pip3] torchsparse==2.1.0
[pip3] torchvision==0.17.0+cu121
[pip3] triton==2.2.0
[conda] numpy                        1.24.4           pypi_0           pypi
[conda] nvidia-cublas-cu12           12.1.3.1         pypi_0           pypi
[conda] nvidia-cuda-cupti-cu12       12.1.105         pypi_0           pypi
[conda] nvidia-cuda-nvrtc-cu12       12.1.105         pypi_0           pypi
[conda] nvidia-cuda-runtime-cu12     12.1.105         pypi_0           pypi
[conda] nvidia-cudnn-cu12            8.9.2.26         pypi_0           pypi
[conda] nvidia-cufft-cu12            11.0.2.54        pypi_0           pypi
[conda] nvidia-curand-cu12           10.3.2.106       pypi_0           pypi
[conda] nvidia-cusolver-cu12         11.4.5.107       pypi_0           pypi
[conda] nvidia-cusparse-cu12         12.1.0.106       pypi_0           pypi
[conda] nvidia-nccl-cu12             2.19.3           pypi_0           pypi
[conda] nvidia-nvjitlink-cu12        12.9.86          pypi_0           pypi
[conda] nvidia-nvtx-cu12             12.1.105         pypi_0           pypi
[conda] pytorch-lightning            2.4.0            pypi_0           pypi
[conda] torch                        2.2.0+cu121      pypi_0           pypi
[conda] torch-cluster                1.6.3+pt22cu121  pypi_0           pypi
[conda] torch-geometric              2.3.0            pypi_0           pypi
[conda] torch-graph-components       0.1.1            pypi_0           pypi
[conda] torch-ransac3d               2.0.0            pypi_0           pypi
[conda] torch-scatter                2.1.2+pt22cu121  pypi_0           pypi
[conda] torch-tb-profiler            0.4.3            pypi_0           pypi
[conda] torchmetrics                 0.11.4           pypi_0           pypi
[conda] torchsparse                  2.1.0            pypi_0           pypi
[conda] torchvision                  0.17.0+cu121     pypi_0           pypi
[conda] triton                       2.2.0            pypi_0           pypi
```

--- a/.gitignore
+++ b/.gitignore
@@ -147,7 +147,9 @@ dmypy.json
 # Lightning-Hydra-Template
 configs/local/default.yaml
 .data/
+/data/
 .logs/
+/logs/
 .env
 .autoenv
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,18 @@ introduced in EZ-SP to a standalone library:
 [torch-graph-components](https://github.com/drprojects/torch-graph-components). 
 This dependency is now installed by `install.sh`.
 
+- Partition training update:
+  - Logging of the number of inter-edges per epoch the model is trained on
+    (`train/n_inter_edge`)
+  - Clearer error or warning messages. Distinct messages for:
+    - error: no edge at all in the batch;
+    - warning: only void edges in the batch;
+    - warning: only intra-edges and no inter-edges in the batch.
+  - The two latter cases no longer raise an error: if a batch is composed of a
+    random crop in an area with only one semantic label, training continues.
+    However, if no inter-edge is found during a whole training epoch, an error
+    is raised, as this likely indicates a bug in the graph or crop definition.
+
 ### Deprecated
 
 ### Fixed

--- a/src/loss/partition_criterion.py
+++ b/src/loss/partition_criterion.py
@@ -1,8 +1,12 @@
 import torch
 from torch import nn
+from typing import Tuple
 
 from src.utils import PartitionOutput, compute_edge_distances_batch
 from src.loss.focal import BinaryFocalLoss
+
+import logging
+log = logging.getLogger(__name__)
 
 __all__ = []
 
@@ -33,7 +37,9 @@ class PartitionCriterion(nn.Module):
         The number of classes.
         
     """
-    
+    INTER_EDGE_LABEL = 0
+    INTRA_EDGE_LABEL = 1
+
     def __init__(self, 
                  loss_function = BinaryFocalLoss(),
                  affinity_temperature: float = 1,
@@ -54,11 +60,16 @@ class PartitionCriterion(nn.Module):
         self.num_classes = num_classes
 
     def forward(self, 
-                partition_output: PartitionOutput):
+                partition_output: PartitionOutput
+                ) -> Tuple[torch.Tensor, PartitionOutput]:
         
-        edge_classification_loss, edge_classification_output = self.edge_classification_loss(partition_output)
+        loss, edge_classification_output, n_inter_edge \
+            = self.edge_classification_loss(partition_output)
+        
+        # Store the number of inter-edges for logging
+        partition_output.n_inter_edge = n_inter_edge
 
-        return edge_classification_loss, partition_output
+        return loss, partition_output
 
     
     def edge_classification_loss(self, partition_output: PartitionOutput):
@@ -78,48 +89,79 @@ class PartitionCriterion(nn.Module):
               For meaningful partition evaluation, see the partition purity metrics logged in 
               `PartitionAndSemanticModule.on_validation_epoch_end`.
         """
-        # Get the taret affinity
+        # A) Get the target edge affinity
         # y is an histogram of the classes of shape (n, C+1), where C is the
         # number of classes and the last column is the void class.
         y = partition_output.y 
         x = partition_output.x
-        
         edge_index = partition_output.edge_index
         
-        # Remove self-loops (which are useless for the loss)
-        mask = edge_index[0] != edge_index[1]
-        edge_index = edge_index[:, mask]
+        if edge_index.numel() == 0:
+            raise ValueError("No edges found in batch.")
         
-        # Taking the argmax assumes the voxels are pure enough.
+        # A.1) Keep only relevant edges
+        
+        # Turn histogram into unique label (by taking the mode per voxel)
         majority_class_count, y = y[:,:self.num_classes].max(dim=1)
         
-        # We discard edges if one of the two nodes is a voxel with only void 
-        # labels (that is, if the count of the majority class is 0)
-        mask_void_voxels = majority_class_count == 0
+        # Remove self-loops (which are useless for the loss)
+        mask_self_loops = edge_index[0] == edge_index[1]
+        edge_index = edge_index[:, ~mask_self_loops]
+        
+        # Discard edges containing a pure void voxel
+        mask_void_voxels = majority_class_count == 0 # all points in the voxel were voids
         mask_void_edges = mask_void_voxels[edge_index[0]] | mask_void_voxels[edge_index[1]]
         edge_index = edge_index[:, ~mask_void_edges]
         
-        # Compute the groundtruth affinity (inter-edges are 0, intra-edges are 1)
-        groundtruth_affinity = (y[edge_index[0]] == y[edge_index[1]]).int()
+        if edge_index.numel() == 0:
+            log.warning("No edges with two non-void nodes found in current batch.\n")
+            return self.fake_edge_classification_loss(y.device)
+        
+        # Compute the target edge affinities 
+        # (inter-edges are 0, intra-edges are 1 - consistent with class 
+        # attributes `INTER_EDGE_LABEL` and `INTRA_EDGE_LABEL`)
+        target_affinity = (y[edge_index[0]] == y[edge_index[1]]).int()
+        n_inter_edge = (target_affinity == self.INTER_EDGE_LABEL).sum().item()
+        
+        if n_inter_edge == 0:
+            log.warning("No inter-edges found in current batch.\n")
+            return self.fake_edge_classification_loss(y.device)
 
-        # Adaptive sampling (only during training)
+        # A.2) Adaptive sampling on edges to balance between inter- and 
+        # intra-edges (ONLY during training)
         if self.training and self.adaptive_sampling_ratio is not None:
-            sampled_indices = self.binary_adaptive_sampling(groundtruth_affinity, 
-                                                            minority_class=0)
+            sampled_indices = self.binary_adaptive_sampling(target_affinity, 
+                                                            minority_class=self.INTER_EDGE_LABEL)
             
-            assert sampled_indices.numel() > 0, \
-                ("No edges left after adaptive sampling. "
-                 "Probably because there were no inter-edges (minority class).")
+            assert sampled_indices.numel() > 0, "No edges left after adaptive sampling."
             
             edge_index = edge_index[:, sampled_indices]
-            groundtruth_affinity = groundtruth_affinity[sampled_indices]
+            target_affinity = target_affinity[sampled_indices]
 
-        # Compute the predicted affinity
+        # B) Predict the edge affinities
         predicted_affinity = self.features_to_edge_affinity(x, edge_index)
-        loss = self.loss_function(predicted_affinity, groundtruth_affinity.bool())
+        loss = self.loss_function(predicted_affinity, target_affinity.bool())
 
-        return loss, ((predicted_affinity>=0.5).int(), groundtruth_affinity)
+        return loss, ((predicted_affinity>=0.5).int(), target_affinity), n_inter_edge
     
+    def fake_edge_classification_loss(self, device):
+        """
+        If for any reason, there are no edges or no inter-edges in the batch to 
+        train on, we return a fake loss and other metrics to avoid errors.
+        
+        This is useful if it happens rarely during one training epoch (e.g., 
+        due to small random crops or specific batch compositions).
+        
+        If during a WHOLE training epoch, no inter-edges are found, then 
+        `PartitionAndSemanticModule.on_train_epoch_end` will raise an error.
+        """
+        # Create a zero loss that requires grad for backward compatibility
+        fake_loss = torch.tensor(0.0, device=device, requires_grad=True)
+        
+        fake_pred = torch.tensor([0], device=device)
+        fake_target = torch.tensor([0], device=device)
+        n_inter_edge = 0
+        return fake_loss, (fake_pred, fake_target), n_inter_edge
     
     def binary_adaptive_sampling(self, y, minority_class=1):
         """

--- a/src/models/semantic.py
+++ b/src/models/semantic.py
@@ -1500,6 +1500,8 @@ class PartitionAndSemanticModule(SemanticSegmentationModule):
         self.val_partition_ooa_best = MaxMetric()
         self.val_partition_omacc_best = MaxMetric()
 
+        self.train_n_inter_edge = SumMetric()
+
     def forward(
             self,
             sample: Union[NAG, Data],
@@ -1570,6 +1572,7 @@ class PartitionAndSemanticModule(SemanticSegmentationModule):
             return super().train_step_update_metrics(loss, output)
         else:
             self.train_partition_loss(loss.detach())
+            self.train_n_inter_edge(output.n_inter_edge)
 
             if self.partition_during_training:
                 y_oracle = output.y_superpoint[:, :self.num_classes].argmax(dim=1)
@@ -1592,7 +1595,36 @@ class PartitionAndSemanticModule(SemanticSegmentationModule):
         if not self.training_partition_stage:
             return super().on_train_epoch_end()
         else:
-
+            n_inter_edge = self.train_n_inter_edge.compute()
+            self.log(
+                    "train/n_inter_edge",
+                    n_inter_edge,
+                    prog_bar=True,
+                    rank_zero_only=True)
+            
+            
+            if n_inter_edge == 0:
+                raise ValueError(
+                    "No inter-edges found during one training epoch.\n\n"
+                    "Possible causes and solutions:\n"
+                    "  1. Random crops are too small (crops are stochastic, therefore, the issue may appear after only a few epochs):\n"
+                    "     - The random crops produced by `SampleRadiusSubgraphs` may only contain "
+                    "intra-edges for a whole epoch.\n"
+                    "     - Solution: Increase `datamodule.sample_graph_r` in your config to get "
+                    "bigger crops.\n"
+                    "     - Alternative: Deactivate crops by setting `datamodule.sample_graph_r=-1` "
+                    "(if GPU memory permits).\n\n"
+                    "  2. Graph definition:\n"
+                    "     - Your dataset and `AdjacencyGraph` graph definition may create only "
+                    "intra-edges or no edges at all.\n"
+                    "     - Solution: Review your graph connectivity parameters (e.g., `k`, `r_max`) "
+                    "in the `AdjacencyGraph` transform."
+                )
+            
+            # Since `self.train_n_inter_edge` is not given to self.log directly,
+            # we need to reset it manually here.
+            self.train_n_inter_edge.reset()
+            
             if self.partition_during_training:
                 self._on_train_epoch_end(
                     cm=self.partition_train_cm,

